KEP-4559 Redesigning Kubelet probes

tssurya opened the pull request:
<!-- 
	Please use the following format when naming your PR
	< Issue Number >:< Issue Description >
	e.g. KEP-000: adding beta graduation criteria
	
	Avoid using phrases like `fixes #NNNN` in the description
	unless the pull request is to change the KEP status to 
	implemented or KEP has been deprecated.
-->

<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- One-line PR description: `Redesigning Kubelet Probes`

<!-- link to the k/enhancements issue -->
- Issue link: https://github.com/kubernetes/enhancements/issues/4559

<!-- other comments or additional information -->
- Other comments:

PS: I am running with what @danwinship did https://github.com/danwinship/enhancements/commit/72333728bb so all credit to him. I am willing to drive this KEP and learn stuff along the way (that is my only role here).

So the main idea here is that the kubelet probes to pods are always expected to work in kubernetes. However when we add network policies (V1) the implicit deny kicks in which means the probes won't work. We went with "NetworkPolicy doesn't apply to Kubelet probes" - this point was originally covered in the NetPol KEP: https://github.com/kubernetes/design-proposals-archive/blob/main/network/network-policy.md#behavior however it never made its way into official documentation and stays as something ecosystem takes as the standard. So all networkpolicy implementations are currently adding allow rules for the probes to work well and creating that hole. SIG-Network-Policy-API would like to tackle this problem better by working with the SIG-Node team

This is also my first PR in enhancement repo, please be patient as I try to get all the stuff correctly done.

danwinship committed:
Redesigning kubelet probes

aojea commented on keps/sig-network/NNNN-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1039 @@
+
+Regardless of what else we do, we will deprecate the
+`TCPSocketAction.Host` and `HTTPGetAction.Host` fields, and allow them
+to be blocked by Pod Security admission.
```
I wonder why this is not the case today, cc @tallclair @liggitt

aojea replied:
related https://github.com/kubernetes/kubernetes/issues/81129

tallclair replied:
Yeah, IMO we should just make this change. I think it is stretching the scope of what PSA covers a bit, but it's not that far.

aojea commented on keps/sig-network/NNNN-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1039 @@
+
+### API Changes to Fix the SSRF
+
+Regardless of what else we do, we will deprecate the
```
can we really deprecate them? I know users that depend on this functionality

danwinship replied:
Do you have examples?

If we know what people are using it for, we can try to come up with an API that allows the "sane" use cases without allowing it to be used for an attack.

aojea replied:
https://docs.tilt.dev/api.html#api.http_get_action
https://github.com/tilt-dev/tilt/blob/3359b56077d1bff23de1d179cb99431eefdd3c35/internal/tiltfile/api/__init__.py#L1385

https://stackoverflow.com/questions/63082309/liveness-probe-of-one-pod-via-another
https://stackoverflow.com/questions/75811572/executing-readiness-httpget-readiness-probe-failed-no-such-host

I remember some user in kubernetes/kubernetes explaining they used the probe to query an external http service but I can not find it ...

danwinship replied:
> https://docs.tilt.dev/api.html#api.http_get_action https://github.com/tilt-dev/tilt/blob/3359b56077d1bff23de1d179cb99431eefdd3c35/internal/tiltfile/api/__init__.py#L1385

That just shows that they're re-exporting the (full) API, not that anyone is actually using it.

> https://stackoverflow.com/questions/63082309/liveness-probe-of-one-pod-via-another

"pod B depends on pod A working so I want pod B's liveness probe to check pod A (via a service externalIP)".

That's a little goofy... Clearly it would be architecturally nicer (if more complicated) to make pod B implement a proper health check itself.

Also, this health check, as written, depends on the existence of the NetworkPolicy hole (since it eventually results in a kubelet-to-pod-A connection) so even if we allowed it, it wouldn't work with AdminNetworkPolicy...

> https://stackoverflow.com/questions/75811572/executing-readiness-httpget-readiness-probe-failed-no-such-host

That's a little more plausible since they're using a service IP that resolves to the pod they're probing, rather than an entirely unrelated pod. This still requires the NetworkPolicy hole though, and this goes back to the "[are probes about server readiness or network connectivity?](https://github.com/kubernetes/enhancements/pull/4558/files#diff-7e38dd798d6b784375dfe279505947c18ea870f56da9fe4d87395c96df27e13dR190-R199)" question

> I remember some user in kubernetes/kubernetes explaining they used the probe to query an external http service but I can not find it ...

That seems even more dubious than the someone-else's-externalIP example.

At any rate, all of these _could_ be implemented via exec probes if it was no longer possible to do them via http probes.

aojea replied:
just to be clear, I'm in favor of this, I also opened the issue some time ago to move the TCP and HTTP probes inside the Pods, and I think that they are different ways of doing this keeping the functionality ... is just that realistically speaking the API and the semantics are v1 and changing them will mean that is a backwards incompatible change :(

MikeZappa87 replied:
For this to work, k8s will need knowledge of the network namespace. It sounds like we want to execute into the netns and run the health checks?

aojea replied:
https://github.com/kubernetes/kubernetes/issues/102613

and there are other tech details in the doc suggested, as reusing existing CRI exec rpc

danwinship replied:
> For this to work, k8s will need knowledge of the network namespace. It sounds like we want to execute into the netns and run the health checks?

There are multiple ideas suggested. And as noted in a comment below, although this KEP was just filed this week, I actually wrote all of it several months ago, so the existing text doesn't consider KNI at all.

aojea commented on keps/sig-network/NNNN-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1039 @@
+2. Kubelet could perform probes inside pods without needing any
+   changes to the runtime, by attaching an ephemeral container to each
+   pod, containing some `kubelet-probe` binary that would run the
+   probes and report the results back to kubelet (by some means). Most
```
I think that this "by some means" is the most important detail here ...

aojea commented on keps/sig-network/NNNN-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1039 @@
+deprecate the existing probe types and create new ones with the new
+semantics.
+
+## Design Details
```
I assume that this is just socializing the problem and opening the discussion, without knowing the design details will be hard to evaluate this KEP

danwinship replied:
Yes. In particular, we need input from SIG Node on which of the ideas here do and don't make sense for kubelet and CRI.

tssurya replied:
yeah that's on me, let me start opening up some discussions here with sig-node

danwinship commented:
> * Issue link:

You need to open an issue to be the "home" for the KEP, and then the number of that issue becomes the KEP number (replacing the "NNNN" in the current branch)

danwinship commented on keps/sig-network/NNNN-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1039 @@
+  turn out that removing the constraint that kubelet needs to be able
+  to talk to all pods would simplify things on Windows?
+
+- Figure out if there are any concerns with non-namespace-based Linux
+  runtimes (kata, kubevirt, etc).
```
It turns out that kata and kubevirt both still create network namespaces and use CNI to set up networking in them, and then they do magic inside the pod network namespace to connect the CNI interface to the VM interface. So there's probably nothing special that needs to be done here.

tssurya replied:
I see let me update this and add this info so that at least we know we reflected upon this.

tssurya replied:
done.

danwinship commented on keps/sig-network/NNNN-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1039 @@
+       report back as part of the pod status whether the pod is live,
+       ready, etc, according to its probes. This might allow for
+       better optimization of probes on the CRI side.
+
```
Since I originally wrote this, we have started talking about [KNI](https://github.com/kubernetes/enhancements/issues/4410), and the ideas here that mention CRI could alternatively be done in terms of KNI. (Assuming that KNI is going to happen in a useful time frame, this would actually make a lot of sense, since it's another case of putting more of the details of the pod network into one place rather than having them be scattered across multiple places.)

danwinship commented on keps/sig-network/NNNN-kubelet-probes/kep.yaml:
```diff
@@ -0,0 +1,39 @@
+title: Pod Probes v2
+kep-number: NNNN
+authors:
+  - "@danwinship"
```
(you should add yourself here)

tssurya replied:
haha will do once I myself contribute something substantial here => I hope to as this KEP evolves since we still need to fill other sections, but as is in current state, let me refrain - that will also motivate me to keep working on my knowledge.

danwinship commented on keps/sig-network/NNNN-kubelet-probes/kep.yaml:
```diff
@@ -0,0 +1,39 @@
+participating-sigs:
+  - sig-node
+status: provisional
+creation-date: 2023-08-23
```
update the date

danwinship commented on keps/sig-network/NNNN-kubelet-probes/kep.yaml:
```diff
@@ -0,0 +1,39 @@
+# The most recent milestone for which work toward delivery of this KEP has been
+# done. This can be the current (upcoming) milestone, if it is being actively
+# worked on.
+latest-milestone: "v1.29"
+
+# The milestone at which this feature was, or is targeted to be, at each stage.
+milestone:
+  alpha: "v1.29"
+  beta: "v1.31"
+  stable: "v1.33"
```
can at least provisionally update these to start in 1.31

tssurya replied:
yeah agreed, changing this to 1.31

danwinship commented:
> We went with "NetworkPolicy doesn't apply to Kubelet probes" and I am not even sure if there is documentation around this (will check and come back to this).

It was in [the original NetworkPolicy design spec](https://github.com/kubernetes/design-proposals-archive/blob/main/network/network-policy.md#behavior) but AFAIK never made it into any official documentation.

tssurya commented:
> > * Issue link:
> 
> You need to open an issue to be the "home" for the KEP, and then the number of that issue becomes the KEP number (replacing the "NNNN" in the current branch)

ah I see thanks Dan, opened https://github.com/kubernetes/enhancements/issues/4559; will link to this KEP

danwinship commented:
/retitle KEP-4559 Redesigning Kubelet probes

tallclair commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+
+## Motivation
+
+### Goals
```
Lifecycle hooks should be included as explicitly in-scope.

tssurya replied:
yep, good point, thanks Tim! will update to include this

thockin reviewed (commented):
Great overview.  Might be good to prioritize the problem statements.  E.g. Maybe SSRF is P0, Performance/timeouts is P1, Dual-stack is P2, NetPol is P4

thockin commented:
We should also clarify and document the behavior in HTTP wrt redirects.

danwinship commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+  (Client.Timeout exceeded while awaiting headers)"). This has been
+  mitigated to some extent ([kubernetes #115143])... it's not clear if
+  this is still considered a problem.
+
```
Is it common to have a kubelet-probing-port which is not also intended as a general serving port? I guess I assumed that people wouldn't bother setting up an HTTP server _just_ so they could answer kubelet probes, and so therefore people would tend to use either (a) an http probe to the actual service being exposed by the pod, or (b) an exec probe

thockin replied:
I have seen cases where there's an HTTP server sidecar which reports a status for a more involved check (periodic exec or converting from some other status mechanism).  Is it a problem to expose that to the network?  Probably not, and we can assert with confidence that if your probe requires a firewall, you are probably doing it wrong.  Still, POLS says thsat 127.0.0.1 would be the pod's own localhost, and it is not.  Yet

tssurya replied:
> I have seen cases where there's an HTTP server sidecar which reports a status for a more involved check (periodic exec or converting from some other status mechanism). Is it a problem to expose that to the network? Probably not, and we can assert with confidence that if your probe requires a firewall, you are probably doing it wrong. Still, POLS says thsat 127.0.0.1 would be the pod's own localhost, and it is not. Yet

What is POLS? (Pod Lifecycle something); sorry if this is a dumb question.. :(
but hold on so IIUC if I give TCPSocketAction.Host == 127.0.0.1 its going to expose the node's localhost and not pod's localhost huh? wow -_-

tssurya replied:
dan told me POLS = Principle of least surprise.

tssurya replied:
I can include this as a point and call this out as well.

danwinship commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+- Continue to backward-compatibly support all existing probes by
+  default (at least for a while).
+
+- Deprecate the [`TCPSocketAction.Host`] and [`HTTPGetAction.Host`]
```
@tallclair also agreed that we should add Pod Security Admission support for blocking these. Does this even need a KEP or can we just call it "fixing https://github.com/kubernetes/kubernetes/issues/99425"?

thockin replied:
Adding new fields usually means a KEP to track gates and so forth.

tssurya replied:
should I open a new KEP issue to track this? OR should it be that we first need to provide a way for users to be able to do the "External services" use case still a.k.a we finish this KEP before opening doors for the deprecation KEP?

danwinship commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+    - Allow administrators to use [Pod Security admission] to block
+      Pods that have probes using those fields.
+
+- Move to a system where TCP, HTTP, and GRPC pod probes are run inside
```
We don't currently support localhost probes though, right?

> may actually need to specify an IP or host or interface?

In general, except for localhost, the user won't know the IP ahead of time.

I feel like support for Multi-Network may need to be a new extension to the probe data once we've defined how Multi-Network works. (Though if we said you specified by interface then that would solve the localhost case right away...)

This also sort of ties into the "probe the non-default IP family" case (in that you want to probe a non-default IP but don't know ahead of time what that IP is).

thockin replied:
> We don't currently support localhost probes though, right?

No, but if I knew that probes ran from within a netns, I might wonder why I can't use localhost :)

I am happy to NOT do this, as long as we choose not to do it, rather than forget to consider it.

danwinship commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+    already more constrained than the TCP and HTTP probes anyway (and
+    I don't have a good suggestion for renaming them).
+
+### Kubelet/Runtime Changes to Fix the NetworkPolicy Hole
```
Well, the network model says that all pods can talk to all other pods too; if you apply NetworkPolicy, you are explicitly rejecting the more-open model. We just need to make sure that it's possible for the users to specify rules that give access to node agents if needed.

ANP already has `nodeSelector` for specifying that you want to allow access from/to the IPs of some/all nodes. We should probably add some way of specifying "the pod's own node".

Note that as specified in the Non-Goals, we are not changing the behavior of v1.NetworkPolicy, for exactly this reason. We're only removing the automatic-same-node-allow from ANP and from "future APIs" (DeveloperNetworkPolicy aka "NetworkPolicy v2").

thockin replied:
The network model assumes no policy (I thought it even states "absent any other policy" or something).  Either way, things that used to work should keep working, as you said

tssurya replied:
The fact that we don't plan to change existing behavior/definition of NetPol and everything works as is is covered in "Non-Goals". Resolving this comment.

danwinship commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+       any probing pod would then be able to connect to any pod in the
+       cluster.
+
+     - Alternatively, the "probing pod" could be a static pod with a
```
Yeah, we've been talking about "pods attached only to a secondary network, not the cluster default pod network" in OpenShift, and one tricky bit is: how does DNS work for them? The current thinking is that pods that "aren't attached to the cluster default pod network" will still actually be attached to the cluster default pod network, just with routing/policy set up to only allow it to be used in very limited ways (like DNS).

But having a special "infra network" attached to every pod would be another way of doing that same thing...

The trick is finding a CIDR range to use; we can't use `169.254.0.0/16` because many existing clusters expect pods to be able to connect to the `169.254.169.254` outside the node. IPv6 link-local addresses would work, but we can't make the infra network be IPv6-only because then we couldn't do health checks for IPv4-only pods over it.

It's still an interesting idea...

aojea replied:
aslo link-local addresses are commonly used by different things to provide this sort of mechanism, if following that path we should make clear what is the range reserved for kubernetes probes ... 

 IPv6 has this concept of address scopes https://datatracker.ietf.org/doc/html/rfc4007#section-6 that solves this problem, as we can create the probe address scope range ... if we just could ensure IPv6 is enable on all clusters, adding this IPv6 capability to pods will certainly solve this problem with very low effort

danwinship commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+Moving pod probes into the pod network namespace will run into the
+standard difficulties with using kernel namespaces from golang code.
+
+The `setns()` syscall in Linux changes the namespace of a thread; this
```
Sure, but containerd and cri-o are both written in Go too, so we're just handing the footguns to them.

(This section is kind of vague but the point is mostly "don't forget about this when deciding which solution is best".)

thockin replied:
Compartmentalizing it is still useful.  Kubelet should not know about netns directly.  CRIs do.  They can do things like open a UNIX socket and fork/exec() a pool of helper processes which can jump between netns'es. to do probes.  It's a little multi-process system of its own, but it's not THAT complicated :)

danwinship commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+what these "sufficiently weird things" would be. One example would be
+having the pod recognize the node IP as a source IP so that it can
+respond differently to kubelet probes than it would to a "real"
+request.) It would be safer (though more annoying) to instead
```
So should we say we definitely want to go with

- Add new probe types (with new semantics and no `.Host` support)
- Deprecate the old probe types
- Allow using PSA to block Pods that use the old probe types

rather than 

- Change the semantics of the existing probe types
- Deprecate the `.Host` fields of the existing probe types
- Allow using PSA to block Pods that use probes with non-empty `.Host` fields

?

(Maybe the first case should also say "block Pods that use the old probes with non-empty `.Host` fields" rather than "block Pods that use the old probes"? I guess I was sort of abusing PSA as a way to help admins force users into the new probe semantics, but that's not really _security_ so maybe it's wrong to use PSA for that. I guess the admin could write their own admission controller if they wanted to fully deprecate the old probes.)

thockin replied:
> Allow using PSA to block Pods that use the old probe types

PSA or simply `ValidatingAdmissionPolicy` - it should solve pretty much all such "don't use this field" use-cases.

If we are going to do probes from in-pod network, then I think yes.  Depending on timeline, we might still want to deprecate `.Host` and recommend `VAP` for admins who want to disallow `.Host` right away.

tssurya replied:
I have started https://github.com/kubernetes/kubernetes/pull/125271 to use PSA for "don't use this field" be careful part.

danwinship commented:
> We should also clarify and document the behavior in HTTP wrt redirects.

Following redirects in general reopens the SSRF problem, so I guess the options are:
- a redirect response is considered success (and is not followed)
- a redirect response is considered failure (and is not followed)
- a redirect to the same host is followed but redirects to off-host are treated as failure (and not followed)

thockin commented:
I think we follow local redirects today but not remote.  I recall this being a CVE or something a long time ago, but I forget the exact resolution.

aojea commented on keps/sig-network/4559-kubelet-probes/README.md:
````diff
@@ -0,0 +1,1037 @@
+<<[/UNRESOLVED]>>
+```
+
+- Continue to backward-compatibly support all existing probes by
````
These use case with NodePort services is interesting 
 https://github.com/kubernetes/kubernetes/issues/89898#issuecomment-2116237787 as it has a lot of network implications

aojea commented:
> I think we follow local redirects today but not remote. I recall this being a CVE or something a long time ago, but I forget the exact resolution. We should document the behavior better.

https://github.com/kubernetes/kubernetes/pull/75416

adrianmoisey commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+  to avoid continuing to have this problem in [AdminNetworkPolicy] and
+  any [future successor to NetworkPolicy].
+
+- **Contraints on Network Architecture Imposed by Pod Probes**: The
```
Typo: ```suggestion
- **Constraints on Network Architecture Imposed by Pod Probes**: The
```

adrianmoisey commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+  This led to the creation of [readiness gates] to provide better pod
+  readiness information.
+
+- **Exec Probe Performance (?)**: [Exec probes are much, much slower than
+  network probes]. Apparently this slowness is a somewhat inevitable
```
I was unaware of this slowness problem.
Is there more context for me to go read up on it? An issue perhaps?

adrianmoisey replied:
Oh, I see below that "Figure out details of exec probe performance problem" is listed

danwinship replied:
The brackets in the text here indicate that that's a link, with the actual URL done "footnote-style" a bit further down in the text. In this case, the link is to a blog post, "[Exec Probes: A Story about Experiment and Relevant Problems](https://medium.com/netcracker/exec-probes-a-story-about-experiment-and-relevant-problems-12de616c0a76)". I don't have much more context than that, other than my guesses in this paragraph. (Though also, I don't think it was ever confirmed that this is a problem with exec probes in all CRI runtimes, or if it's basically just a containerd bug.)

tssurya replied:
I can add a note to that effect that we have not had any specific issues opened in the past around this but listing the problem here based on the post.

aojea replied:
we need someone from Node to evaluate this or we have the risk of operating with biased information, @samuelkarp can you help us to understand better this performance problem with Exec probes? is this a general and unsolvable problem?

mikebrow replied:
exec with a new task is not as expensive as running a container..but still, it's not an inexpensive operation either.. and the usual problem is there are a plethora of exec probes that are not cycle optimized because of the default timeouts and repeat duration values.. And when the system gets stressed the probe failures just make it worse.

We could probably figure out more optimizations... sure that's worthy of trying even if just for non-probe exec. Would be interesting to see if there is a difference between crio/containerd or any other improvements since the last time we checked. 

Though, I held out hope for a different probe api that did not lean on/overload exec and the streaming api.

tssurya commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+There are several problems with the existing Pod liveness, readiness,
+and startup probes:
+
+- **Probe-based Blind SSRF Attacks**: The current definition of TCP and
```
I read through this issue and learnt a ton of new things!
So I think from all that discussion it was narrowed down to either use [PSA](https://kubernetes.io/docs/concepts/security/pod-security-admission/) to restrict users from being able to do this OR go the Kubelet/CRI way and let the probes happen in pod network namespace as potential solutions?

> IMO this "feature" was a total mistake, and we should seek to kill it.

does this mean totally deprecate the `Host` API?

> The workaround is a sidecar helper which does the external probe and converts it to a simple HTTP probe.

as in the pod applications that rely on external connections could use the sidecar to see if its a success and then the actual probe can be reliant on that so we don't want kubelet to be the probing external services instead the pod itself can?

tssurya replied:
This has been discussed extensively here and on the issue and on other comments. The plan forward is:

1. use PSA to block users from using these via versioned profiles (https://github.com/kubernetes/kubernetes/pull/125271)
2. Have a new KEP or use this same KEP to deprecate the probes and lifecycle hooks totally
3. Use this KEP to design how the new probes will look like.

tssurya commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+  IPv6 probe in an IPv4-primary dual-stack cluster. ([kubernetes
+  #101324])
+
+- **Unclear Semantics for Network Probes**: It was never defined
```
Tried to clarify the intention versus interpretation better.

tssurya commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+
+### API Changes to Fix the SSRF
+
+Regardless of what else we do, we will deprecate the
```
changed to include points 1/2 for now; Is adding a metric to inform users of deprecated fields usage something we do usually? If so I could add that one too

tssurya commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+      rather than by modifying the semantics of the existing probes,
+      then it should be possible to block the old probe types via Pod
+      Security admission, so that administrators can create
+      AdminNetworkPolicies that don't include the kubelet probe hole.
```
this is not possible today in AdminNetworkPolicy API but we might add support to block ingress from nodes in future if a use case presents itself for that. In that case this point is valid, I'll repush to clarify that.

tssurya replied:
removed this point after discussing with @danwinship and he agreed. resolving.

tssurya commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+    already more constrained than the TCP and HTTP probes anyway (and
+    I don't have a good suggestion for renaming them).
+
+### Kubelet/Runtime Changes to Fix the NetworkPolicy Hole
```
@mrunalp : Could you PTAL this section and let us know your thoughts? (I think you and @aojea had talked about this at some point in the past)

aojea commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+2. Kubelet could perform probes inside pods without needing any
+   changes to the runtime, by attaching an ephemeral container to each
+   pod, containing some `kubelet-probe` binary that would run the
+   probes and report the results back to kubelet (by some means). Most
```
the by some means here is the key

aojea commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+   container to the pod at construction time, via CRI calls, without
+   exposing the container to the kubernetes API, like it does with the
+   pause / sandbox / infrastructure container. (In fact, the probes
+   could even be done as part of the existing pause / sandbox /
```
this is going to be removed in containerd 2.0, pause sandbox is an implementation detail and we can not depend on it

aojea commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+   than being a new ephemeral container attached each time kubelet
+   needed to perform a probe. Kubelet could perhaps attach such a
+   container to the pod at construction time, via CRI calls, without
+   exposing the container to the kubernetes API, like it does with the
```
there can be a mismatch on resources consumption also, and the whole lifecycle of the pod will be compromised, if something happens with the probe container do we block the pod? how can user debug something if is hidden? it has a lot of implications we need to consider carefully

aojea commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+   could even be done as part of the existing pause / sandbox /
+   infrastructure container.)
+
+3. If we can improve the performance of exec probes, then kubelet
```
This topic needs some assessment from Node people https://github.com/kubernetes/enhancements/pull/4558/files#r1620722068

aojea commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+
+#### UNRESOLVED Summary discussions
+
+- Figure out details of exec probe performance problem.
+
+- Figure out if network probe performance is still a problem.
```
@mikebrow you may be interested in this KEP, it may help to  implement subseconds probes https://github.com/kubernetes/enhancements/pull/3067

aojea commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+
+- Allow dual-stack / wrong-single-stack probing, either based on
+  information in the probe definition (e.g. `ipFamily: IPv6` or
+  `ipFamilyPolicy: RequireDualStack`) or by always doing "[Happy
```
I would not say that happy eyeballs is a successful protocol , it also had to go through a new version https://datatracker.ietf.org/doc/html/rfc8305 and is very unreliable and causes a lot of issues in production environments

danwinship replied:
(Problems with actual Happy Eyeballs are not relevant. I just meant the general idea of "try both IPs and consider the probe to have succeeded if either works". We don't have to worry about the things actual Happy Eyeballs does (like the possibility that the packets will have to traverse an IPv4-only network on their way from a dual-stack client network to a dual-stack endpoint network).)

aojea replied:
I do not think that a probe that has an OR on the protocol can be reliable, as it can flap, sometimes IPv4 works , sometimes IPv6 works, the probe is always ok ... probes should be reliable and assert on the condition

tssurya commented:
The recent push addresses comments till `Kubelet/Runtime Changes to Fix the NetworkPolicy Hole` -> I am continuing to addressing comments from that onwards.

thockin commented:
Running up against the clock - how much of this is slated for 1.31?

tssurya commented:
> Running up against the clock - how much of this is slated for 1.31?

I would say only the PSA bits where we want to restrict users can land for 1.31, I have added an item to ask sig-auth if that need's a separate KEP.
Outside of that the main solution around "if doing CRI API is the final way to go" or how about the other options - still needs more reviews/+1's so might not happen in the 1.31 time frame. cc @danwinship lmk if you think otherwise.

tssurya commented:
/remove-lifecycle stale

bowei commented:
/CC @bowei

tssurya commented:
So coming back to this after a long break - I want to try to keep pushing at this slowly this year...

Plan is that in the 1.34 frame I want to get the first smaller KEP in sig-auth: https://github.com/kubernetes/enhancements/pull/4942 merged first which will use PSA to block .host field

The last update for this work stands at the fact that @danwinship and I had presented at the SIG-Node meeting on [5th November 2024](https://www.youtube.com/watch?v=1u_yKruHeZU&list=PL69nYSiGNLP1wJPj5DYWXjiArF-MJ5fNG&index=14). I want to summarize that discussion here before I include a new commit to add those bits into the PR:

1. The idea of probing pods (one probing pod per node), i.e using CNI pods to do the probes was well-received. @dchen1107 had mentioned that having a proper pod doing the probes sounds better than what we have today - its more manageable than the exec style we have today. Also the cost of doing probes then falls on the user and complexity pushed to pod from infra-level. Overhead is part of the system. At the application level usage is unpredictable.
  i. @mrunalp had also suggested an additional container within each pod v/s a probing pod? - this was mainly let's say so that the probing pod still have the same problem which is the fact that is has to be reachable to all pods to be able to do the probing which doesn't solve the netpol loophole
  ii. @danwinship raised the idea of using a single AdminNetworkPolicy to allow this probing podIP to reach all pods on the cluster to be able to do the probes -> that's more official than having to poke holes in the cluster like we do today which makes anything on the node reach the pods for probes to work with policies?

2. Another point raised was the concern around deprecating existing probes in favor of the new probe types -> It was mentioned instead if the approach to keep both probe types and users can choose which one to go with is is better and less scarier -> also along the lines of the behaviour we have today example of ensuring the port is open when done at the host level v/s moving this to pod level probing; also how load balancers use the readiness probes to ensure pod is ready to accept connections
  i. The KEP currently is angling towards deprecating existing and going with new.. but this is still unresolved conversation open for debate.
  ii. but the current probes have issues like -> in the case of load balancer knowing the pod is ready to accept connections but with network policies here actually the pod is not ready to accept any other connections apart from kubelet checks maybe? So how do we clearly state that?
  iii. Properties of the new probes v/s the old probes and performance comparisons will need to be done as well before going down that road

3.  File Based Probes -> instead of CRI based
     i. check if a file exists and that's a probe type
     ii. more performant efficient but what if the file cleanup fails or something

4. RE: There was a brief discussion around also perhaps having CRI method for probes - so making CRI API call for probes behave like TCP Proxy - so instead of kubelet doing the probes, offloading this to the CRI and runtime level as mentioned as an option in this KEP
   i. Is there already an existing proxy CRI method today? -> YES this is how exec and attach work today (bunch of proxies between API/Kubelet/CRI and then CRI exposes a url that the client connects to) -> so the architecture for proxy exists we could expand on that hook perhaps? Port forwarding also works this way?

ACTION from the meeting was to update the KEP with the above discussion points and to take that back to the sig-node community for a re-review.

cc @mrunalp : @danwinship and I are doing a contributor summit talk around this KEP, if you and other sig-node community members will be at london maintainer summit please stop by: https://sched.co/1tj91
I want to see if we can somewhat converge on a preferred way?

I'll add a new commit to this KEP to include those details.

tssurya commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,1037 @@
+   a UDP probe) would require adding support to both `Pod`/kubelet and
+   to CRI.
+
+     - If CRI is going to need to understand that much of the
```
I think that is a fair point..

vinaykul commented:
/cc

tssurya committed:
Update the KEP based on KubeCON London feedback

Signed-off-by: Surya Seetharaman <suryaseetharaman.9@gmail.com>

tssurya commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -213,35 +217,25 @@ UNRESOLVED section in the Summary.
-      Security admission, so that administrators can create
-      NetworkPolicies that don't include the kubelet probe hole.
-
 - Allow dual-stack / wrong-single-stack probing, either based on
```
We could move this into future goals and outline the fact that we should have expandable changes that don't lock us in in case in future we want to do this work

tssurya commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -213,35 +217,25 @@ UNRESOLVED section in the Summary.
 
 ### Non-Goals
 
 - Adding entirely-new probe types.
```
Adding entirely new probe types that are not replacement for the existing probe types

tssurya commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -213,35 +217,25 @@ UNRESOLVED section in the Summary.
 
 - Adding entirely-new probe types.
 
+  - A potential future goal could be to have a new probe that explicitly
```
Move to future goal

tssurya commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -296,18 +291,62 @@ Beyond that, there are two (and a half) possible plans:
+
+["Redesigning Kublet Probes"]: https://www.youtube.com/watch?v=-pefxR_SwaY
+
+#### Alternatives
```
move this to the main alternatives section

tssurya commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -265,13 +259,14 @@ nitty-gritty.
 
 ### API Changes to Fix the SSRF
```
add another section for https://github.com/kubernetes/kubernetes/issues/131183
or sub divide the API section to include ports as a point for consideration and ensure we fix this issue

danwinship committed:
updates

tssurya committed:
Merge pull request #1 from danwinship/kubelet-probes-updates

KEP-4559 updates

tssurya commented:
We need to consider a more cleaner definition for when a pod is ready and we need to add a new type of pod probe to dictate when the networking is ready for the pod = pod readiness gate

bowei commented:
/lifecycle frozen

SergeyKanzhelev commented:
/sig node

(to make sure sig node reviewed if this will be moving forward)

aojea commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,992 @@
+
+        - Probe is done from within the pod network namespace, so does
+          not guarantee that the network plugin has connected the pod
+          to the pod network.
```
> For example, we could run a prober in the cluster and probe pods from a couple different nodes

I've seen that before and it tends to cause the opposite effect, you aim to get more signal and rather give it you more noise since now your inter-node connectivity impact the state ... it also add more load to the cluster and wastes network resources like conntrack entries and tcp sockets , remember https://github.com/kubernetes/kubernetes/pull/115143

vinaykul commented:
This KEP is interesting because current probe design affects our security posture as well. We use short-lived tokens (generated by eBay's homegrown TrustFabric security platform, spec: https://github.com/eBay/trustfabric) for S2S interactions (micro-segmentation). And we also need to make an exception for probes in the framework token validator or the envoy sidecar token validation filter. For a payments processing company like us, this increases the audit scope due to increased attack surface, which directly translates to increased costs for proving compliance with regulatory standards such as PCI-DSS, NYDFS etc. 

At London maintainer summit, I briefly shared another potential approach - a bit outside the box - of using heartbeats as an optional/alternate mechanism to complement probes. At a very high-level, the application, in good faith, internally performs periodic probing, ideally exercising the same codepaths a kubelet probe would, and if successful, it emits a heartbeat. This would give us the same signal as a probe, but without lowering security posture, and it should address most if not all of the issues identified in this KEP. While this approach comes with its own set of challenges and complexities (especially for kubelet, with many details to be worked out), there's a precedent for it in node heartbeats/kube-node-lease which has worked well. I'm just curious if such an approach was considered in early days of the project before the probe approach was adopted. If it was discussed and ruled out, it would be good to know the reasons.

SergeyKanzhelev commented:
> and if successful, it emits a heartbeat.

Yes, having the low overhead, secure channel from container to kubelet will be great. Container may want to do more than just health reporting eventually: https://github.com/kubernetes/enhancements/issues/5309. The conversation is basically to agree on what this channel my be.

> And we also need to make an exception for probes in the framework token validator or the envoy sidecar token validation filter.

Have you tried exec probes? Do they pose other issues? Or purely performance problems?

vinaykul commented:
> > and if successful, it emits a heartbeat.
> 
> Yes, having the low overhead, secure channel from container to kubelet will be great. Container may want to do more than just health reporting eventually: #5309. The conversation is basically to agree on what this channel my be.
> 
> > And we also need to make an exception for probes in the framework token validator or the envoy sidecar token validation filter.
> 
> Have you tried exec probes? Do they pose other issues? Or purely performance problems?

KEP https://github.com/kubernetes/enhancements/issues/5309 looks interesting, appears to have similar motivation as [dynamic containers](https://docs.google.com/document/d/1J8Aq0XzN8BiNdWHXSEGA1Xw2nXcZRSKTMoi-tNh7FTc/edit?tab=t.0) approach. I’ll take a closer look.

Exec probes are used but in a limited way - typically for startup or for containers that don't serve traffic. We could use them for http health checks but it won't really reduce the audit scope - threat model already assumes that all bets are off if node is compromised; mitigation is compensating controls and RED analysis. Also, shipping curl binary in the container for exec probe bloats image size and only adds one more thing to explain to auditors.

lmktfy commented on keps/sig-network/4559-kubelet-probes/README.md:
```diff
@@ -0,0 +1,992 @@
+# KEP-4559: Pod ProbeHandler/LifecycleHandler v2
```
This title doesn't match the title of enhancement issue 4559. Let's reconcile those to match. 

Try to explain the improvement (more than the motivation).
